app/pages/learn/01_tutorial/04_mastering-the-api/02_invoke/00_methodhandle.md

```diff
@@ -0,0 +1,531 @@
+    System.out.println("inside finally block");
+    return result;
+}
+```
+
+We can wrap the method handle from previous example inside the try-finally block as follows:
+
+```java
+MethodHandle cleanupMethod = lookup.findStatic(Example.class, "cleanupMethod",
+        MethodType.methodType(int.class, IllegalArgumentException.class, int.class, String.class));
+
+MethodHandle wrappedWithFinally = MethodHandles.tryFinally(wrapped, cleanupMethod);
+
+System.out.println(wrappedWithFinally.invoke("valid")); // outputs "inside finally block 1"
+System.out.println(wrappedWithFinally.invoke("invalid")); // outputs "inside finally block 0"
+```
+
+<a id="transformations">&nbsp;</a>
+## Method handle transformations
+As seen from previous examples, method handles can encapsulate more behavior than simply pointing to an underlying method. 
+We can obtain **adapter** method handles, which wrap original method handles to add certain behaviors such as argument reordering, pre-inserting, or filtering of the return values.
+
+Let's take a look at a couple of such transformations.
+
+### Type transformation
+Method handle's type can be adapted to a new type using the `asType` method. If such type conversion is impossible, we will get a `WrongMethodTypeException`.
+Remember, when we apply transformations, we actually have two method handles, where the original method handle is wrapped into some extra logic.
+In this case, the wrapper will take in the arguments and try to convert them to match the original method handle's arguments. Once the original method handle
+does its job and returns a result, the wrapper will attempt to cast this result to the given type.
+
+Assume we have a `test` method that accepts an `Object` and returns a `String`. We can adapt such a method to accept a more specific argument type, such as `String`:
+
+```java
+MethodHandle originalMethodHandle = lookup.findStatic(Adapters.class, "test", 
+        MethodType.methodType(String.class, Object.class));
+MethodHandle adapter = originalMethodHandle.asType(
+        MethodType.methodType(String.class, String.class));
+String originalResult = (String) originalMethodHandle.invoke(111); // works
+String adapterResult = (String) adapter.invoke("aaaaaa"); // works
+adapterResult = (String) adapter.invoke(111); // fails
+```
+
+In fact, each time we use `invoke` on a `MethodHandle`, the first thing that happens is an `asType` call. 
+`invoke` accepts and returns `Object`s, which are then attempted to be converted to more specific types. 
+These specific types are derived from our code, i.e., the exact values that we pass as arguments and the type that we cast our return value to. 
+Once the types are successfully converted, the `invokeExact` method is then called for these specific types.
+
+### Permute arguments
+To obtain an adapter method handle with reordered arguments, we can use `MethodHandles.permuteArguments`.
+
+For example, this is an original method:
```

Comment:
Hey Nataliia,

Can you please elaborate on what you mean by _original method_?